### PR TITLE
documents: add isFiction

### LINKED ISSUE #3356: Documents: add a field to flag fiction documents
### How it works
The specs for differentiating fiction and non-fiction subject facets were incorrect. As of now, we have two different facets that are populated according to a specific subject field in the document.  `FICTIONS_TERMS = ['Fictions', 'Films de fiction']`. This behaviour is not needed since the "Fiction" definition applies to a document and not to subjects. 

### Improvement suggestion
- Add a boolean field in the document isFiction, it is not required.
- Adapt dojson mappings to populate this field on import (MARC21):
    - if MARC `008/33` == `1 `, `d`, `f`, `j` or `p` : isFiction = True
    - elif MARC `008/33` =='0', 'e', 'h', 'i', 's' isFiction = False
    - else isFiction empty
- Adapt dojson mappings to populate this field on import (UNIMARC):
    - if MARC `105/11` == a, b, f, g, i : isFiction = True
    - elif MARC `105/11` == c, d, e, h, y isFiction = False
    - else isFiction empty
- Delete the differentiation for subjects facet, keep only one subject facet
- Add a filter with facet (Fiction) to filter fiction/non-fiction in pro and public UI (document search)
- Check ebook import to see if this information is available and adapt the ebook import. If not available, do not set `isFiction` for imported ebooks.

#### Existing data migration

Write a migration script to set existing documents

- IF "Fictions" OR "Film de fiction" in genreForm: isFiction = True
- ELIF `_exists_:subjects` AND NOT `harvested`: isFiction = False
- ELSE, do nothing (do not input isFiction field)

### PR DESCRIPTION
**Pull request by rerowep**
* Merges alambic 8d97be2c8ad6 and 710021979fe.
* Closes #3356.

### CODE REVIEW

File: rero_ils/config.py [hunk 1]
```diff
@@ -1818,28 +1789,14 @@ def _(x):
...
-            ),
-            _('subject_no_fiction'): and_i18n_term_filter(
-                'facet_subject',
-                must_not=[{
-                    'name_or_query': 'terms',
-                    # TODO: If we have translated concepts we have to addapt this filter.
-                    'facet_genre_form_en': FICTIONS_TERMS}
-                ]
-            ),
+            _('subject_no_fiction'): and_i18n_term_filter('facet_subject'),
```
Comment (PascalRepond) [inline rero_ils/config.py:1792]: Rename this to "subject"?

File: rero_ils/config.py [hunk 2]
```diff
@@ -1818,28 +1789,14 @@ def _(x):
...
             # This filter is used with timestamp
             _('acquisition'): acquisition_filter(),
             # This filter is only used for constructed queries
             # --> Ex: &new_acquisition=2020-01-01:2021-01-01
             _('new_acquisition'): acquisition_filter(),
-            _('identifiers'): nested_identified_filter()
+            _('identifiers'): nested_identified_filter(),
+            _('subject_fiction'): bool_filter('isFiction'),
```
Comment (PascalRepond) [inline rero_ils/config.py:1799]: Rename this to "fiction"?

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/kul/model.py [hunk 3]
```diff
@@ -57,6 +57,9 @@ def marc21_to_language(self, key, value):
     languages: 008 and 041 [$a, repetitive]
     """
     language = do_language(self, marc21)
+    # is fiction
+    if value[33] in ['1', 'd', 'f']:
+        self['isFiction'] = True
```
Comment (PascalRepond) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/kul/model.py:62]: I think I would put:
```python
if value[33] in ['1', 'd', 'f', 'j', 'p']
    self['isFiction'] = True
elif value[33] in ['0', 'e', 'h', 'i', 's']
    self['isFiction'] = False
```

File: rero_ils/modules/documents/dojson/contrib/unimarctojson/model.py [hunk 4]
```diff
@@ -1182,3 +1182,12 @@ def marc21_to_electronicLocator_from_field_856(self, key, value):
     return (
         {'url': value.get('u'), 'type': 'resource'} if value.get('u') else None
     )
+
+
+@unimarc.over('isFiction', '^105..')
+@utils.ignore_value
+def marc21_to_isFiction(self, key, value):
```
Comment (PascalRepond) [inline rero_ils/modules/documents/dojson/contrib/unimarctojson/model.py:1189]: unimarc_to_is_Fiction?

File: rero_ils/modules/documents/dojson/contrib/unimarctojson/model.py [hunk 5]
```diff
@@ -1182,3 +1182,12 @@ def marc21_to_electronicLocator_from_field_856(self, key, value):
...
+    """Get isFiction from field 105 $a 11."""
+    if subfield_a := value.get('a'):
```
Comment (PascalRepond) [inline rero_ils/modules/documents/dojson/contrib/unimarctojson/model.py:1191]: codes for fiction=True : a, b, f, g, i
codes for fiction=False : c, d, e, h, y

**Review by Garfield-fr — APPROVED** [review]

File: rero_ils/modules/documents/mappings/v7/documents/document-v0.0.1.json [hunk 6]
```diff
@@ -2116,6 +2116,9 @@
       "status": {
         "type": "keyword"
       },
+      "isFiction": {
```
Comment (jma) [inline rero_ils/modules/documents/mappings/v7/documents/document-v0.0.1.json:2119]: isFiction or `fiction` , harvested and fiction should have the same naming or nomenclature.

**Review by jma — COMMENTED** [review]
Can you add some tests on the REST API? Thanks.

File: rero_ils/alembic/bd78d77eb7e3_isfiction.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def upgrade():
+    """Upgrade database."""
+    FICTIONS_TERMS = ['Fictions', 'Films de fiction']
+    query = DocumentsSearch() \
+        .filter('terms', facet_genre_form_en=FICTIONS_TERMS)
+    click.echo(f'Add isFiction=true to documents: {query.count()}')
+    for idx, hit in enumerate(query.source('pid').scan(), 1):
+        if doc := Document.get_record_by_pid(hit.pid):
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/alembic/bd78d77eb7e3_isfiction.py:38]: Yes I could but then the line is to long.

File: rero_ils/modules/documents/mappings/v7/documents/document-v0.0.1.json (hunk 6 above)
Comment (rerowep) [reply rero_ils/modules/documents/mappings/v7/documents/document-v0.0.1.json:2119]: changed to fiction

File: rero_ils/alembic/bd78d77eb7e3_isfiction.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    LOGGER.info(f'Add fiction=true to documents: {query.count()}')
+    for idx, hit in enumerate(query.source('pid').scan(), 1):
+        if (doc := Document.get_record_by_pid(hit.pid)):
+            if not doc.get('fiction'):
+                LOGGER.info(f'{idx:<10} {hit.pid:<10} add fiction=true')
+                doc['fiction'] = True
+                doc.update(data=doc, dbcommit=True, reindex=True)
+    query = DocumentsSearch() \
+        .filter('exists', field='genreForm') \
+        .exclude('term', harvested=True)
```
Comment (PascalRepond) [inline rero_ils/alembic/bd78d77eb7e3_isfiction.py:47]:
```python
    query = DocumentsSearch() \
        .exclude('term', harvested=True)
        .exclude('terms', facet_genre_form_en=FICTIONS_TERMS) \
        .filter('exists', field='subjects') \
 ```

**Review by PascalRepond — CHANGES_REQUESTED** [review]

File: rero_ils/alembic/bd78d77eb7e3_isfiction.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            if not doc.get('fiction'):
+                LOGGER.info(f'{idx:<10} {hit.pid:<10} add fiction=false')
+                doc['fiction'] = False
+                doc.update(data=doc, dbcommit=True, reindex=True)
+
+
+def downgrade():
+    """Downgrade database."""
+    query = DocumentsSearch() \
+        .filter('term', fiction=True)
```
Comment (PascalRepond) [inline rero_ils/alembic/bd78d77eb7e3_isfiction.py:61]: We should remove the field in all cases. 
Should it be ```.filter('exists', field=fiction)``` ?

**Review by PascalRepond — APPROVED** [review]

File: rero_ils/alembic/bd78d77eb7e3_isfiction.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2022 RERO
```
Comment (lauren-d) [inline rero_ils/alembic/bd78d77eb7e3_isfiction.py:4]: `Copyright (C) 2024 RERO`

File: rero_ils/alembic/bd78d77eb7e3_isfiction.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def upgrade():
+    """Upgrade database."""
+    FICTIONS_TERMS = ['Fictions', 'Films de fiction']
+    query = DocumentsSearch() \
+        .filter('terms', facet_genre_form_en=FICTIONS_TERMS)
+    LOGGER.info(f'Add fiction=true to documents: {query.count()}')
+    for idx, hit in enumerate(query.source('pid').scan(), 1):
+        if (doc := Document.get_record_by_pid(hit.pid)):
```
Comment (lauren-d) [inline rero_ils/alembic/bd78d77eb7e3_isfiction.py:40]: You can rewrite like this
```python
query = DocumentsSearch() \
        .filter('terms', facet_genre_form_en=FICTIONS_TERMS)
LOGGER.info(f'Add fiction=true to documents: {query.count()}')
ids = [hit.meta.id in hit for query.source().scan()]
LOGGER.info(f'Add fiction=true to documents: {query.count()}')
for idx, _id in enumerate(ids, 1):
  if (doc := Document.get_record(_id)):
```
in this way you prevent elasticsearch timeout and you save many database calls.
for reminder: with `get_record_by_pid` does 2 calls to retreive record and `get_record` does one call.

File: rero_ils/alembic/bd78d77eb7e3_isfiction.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+                LOGGER.info(f'{idx:<10} {hit.pid:<10} add fiction=true')
+                doc['fiction'] = True
+                doc.update(data=doc, dbcommit=True, reindex=True)
+    query = DocumentsSearch() \
+        .exclude('term', harvested=True) \
+        .exclude('terms', facet_genre_form_en=FICTIONS_TERMS) \
+        .filter('exists', field='subjects')
+    LOGGER.info(f'Add fiction=False to documents: {query.count()}')
+    for idx, hit in enumerate(query.source('pid').scan(), 1):
+        if doc := Document.get_record_by_pid(hit.pid):
```
Comment (lauren-d) [inline rero_ils/alembic/bd78d77eb7e3_isfiction.py:51]: same as before

File: rero_ils/alembic/bd78d77eb7e3_isfiction.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                doc.update(data=doc, dbcommit=True, reindex=True)
+
+
+def downgrade():
+    """Downgrade database."""
+    query = DocumentsSearch() \
+        .filter('exists', field='fiction')
+    LOGGER.info(f'Remove fiction from documents: {query.count()}')
+    for idx, hit in enumerate(query.source('pid').scan(), 1):
+        if doc := Document.get_record_by_pid(hit.pid):
```
Comment (lauren-d) [inline rero_ils/alembic/bd78d77eb7e3_isfiction.py:64]: same as before

**Review by lauren-d — APPROVED** [review]

**Review by jma — APPROVED** [review]

**Pull request merged**